mteb/evaluation/MTEB.py

```diff
@@ -469,55 +480,74 @@ def run(
-                )
-                missing_splits = self._get_missing_splits(
-                    existing_results, task_eval_splits
-                )
-
-                if not missing_splits and existing_results:
-                    evaluation_results.append(existing_results)
+            # Unified call to get missing splits and subsets
+            missing_evaluations = self._get_missing_evaluations(
+                existing_results,
+                task_eval_splits,
+                task_subsets,
+                eval_subsets,
+            )
 
-                    # no splits are evaluated.
-                    self.last_evaluated_splits[task.metadata.name] = []
-                    del self.tasks[0]
-                    continue
+            # Determine final splits to run
+            final_splits_to_run = []
+            # We need to run any split that is fully missing or has missing subsets
+            for sp, info in missing_evaluations.items():
+                if info["whole_split_missing"] or info["missing_subsets"]:
+                    final_splits_to_run.append(sp)
 
-                if missing_splits:
+            # If no splits need to be run and results exist, skip
+            if not final_splits_to_run:
+                if existing_results is not None:
+                    evaluation_results.append(existing_results)
+                else:
                     logger.info(
-                        f"Running evaluation for missing splits: {missing_splits}"
+                        f"No splits to evaluate for {task.metadata.name}. Skipping evaluation."
                     )
+                self.last_evaluated_splits[task.metadata.name] = []
+                del self.tasks[0]
+                continue
 
             try:
                 task.check_if_dataset_is_superseded()
                 task.load_data(eval_splits=task_eval_splits, **kwargs)
 
-                # run evaluation
                 task_results = {}
                 evaluation_time = 0
                 kg_co2_emissions: int | None = 0 if co2_tracker else None
 
                 self.last_evaluated_splits[task.metadata.name] = []
 
-                for split in missing_splits:
+                for split in final_splits_to_run:
+                    info = missing_evaluations[split]
+
+                    # Determine subsets to run for this split
+                    # If the whole split is missing, run all required subsets
+                    # If only some subsets are missing, run only those
+                    if info["whole_split_missing"]:
+                        if task_subsets is not None:
+                            subsets_to_run = info["missing_subsets"]
+                        else:
+                            subsets_to_run = ["default"]
+                    else:
+                        # partial subsets missing
+                        subsets_to_run = info["missing_subsets"]
```

Comment:
Can be simplified to reduce mental load.
```suggestion
                    subsets_to_run = info["missing_subsets"]
                    if info["whole_split_missing"] and task_subsets is None:
                        subsets_to_run = ["default"]
```